### ISSUE TITLE
Allow for accessing target name in DSL

### ISSUE THREAD

**Issue by kendonB** [ropensci/drake#827]
I would like to use `transform = map(.)` to output many file targets but haven't been able to figure out how to do it or whether it's currently possible. If it's yet to be implemented, you could allow for referencing the name of the target using something like `.target_name`:

``` r
library(drake)
drake_plan(
  data = target(
    get_data(param),
    transform = map(param = c(1, 2))
  ),
  results_one_file = target(
    save_file(get_results(data), file_out(paste0(.target_name, ".filetype"))),
    transform = map(data, .id = FALSE)
  )
)
#> # A tibble: 4 x 2
#>   target           command                                                 
#>   <chr>            <expr>                                                  
#> 1 data_1           get_data(1)                                            ~
#> 2 data_2           get_data(2)                                            ~
#> results_one_file   save_file(get_results(data_1), file_out(paste0(.target_name,      ".filetype")))
#> results_one_file_2 save_file(get_results(data_2), file_out(paste0(.target_name,      ".filetype")))
```

<sup>Created on 2019-04-08 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1.9000)</sup>

which might result in the last two lines being:

```r
#> results_one_file   save_file(get_results(data_1), file_out(paste0("results_one_file", ".filetype")))
#> results_one_file_2 save_file(get_results(data_2), file_out(paste0("results_one_file_2", ".filetype")))
```

**Comment by wlandau** [comment]
I was initially resistant to making a special exception for this, but in light of https://github.com/sol-eng/tensorflow-w-r/blob/master/workflow/tensorflow-drake.Rmd#L283, I am more open to it. The following functionality is trivially easy to implement.

``` r
library(drake)
as_chr <- function(x) {
  deparse(substitute(x))
}
plan <- drake_plan(
  data = target(
    get_data(param),
    transform = map(param = c(123, 456))
  ),
  model = target(
    save_model_hdf5(fit_model(data), file_out(!!sprintf("%s.h5", .id_chr))),
    transform = map(data, .id = param)
  ),
  result = target(
    predict(load_model_hdf5(file_in(!!sprintf("%s.h5", as_chr(model))))),
    transform = map(model, .id = param)
  )
)
drake_plan_source(plan)
#> drake_plan(
#>   data_123 = get_data(123),
#>   data_456 = get_data(456),
#>   model_123 = save_model_hdf5(fit_model(data_123), file_out("model_123.h5")),
#>   model_456 = save_model_hdf5(fit_model(data_456), file_out("model_456.h5")),
#>   result_123 = predict(load_model_hdf5(file_in("model_123.h5"))),
#>   result_456 = predict(load_model_hdf5(file_in("model_456.h5")))
#> )
```

<sup>Created on 2019-04-08 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

The custom `as_chr()` function is a little snag. I may make it part of `drake` if the `rlang` does not already have a function like it.

**Comment by wlandau** [comment]
Implemented in 29d1214e389df370367fda1202b62fab9c7d53e6.

**Comment by wlandau** [comment]
Note that `.id_chr` is a character string, not a symbol. This has 2 advantages:

1. You do not need to worry about converting symbols to strings when you declare output files.
2. With symbols, we risk accidental cycles in the DAG.
